group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,454 @@
+                throw new IllegalArgumentException("GroupMetadataManager cannot be null");
+            }
+
+            return new OffsetMetadataManager(
+                snapshotRegistry,
+                logContext,
+                time,
+                metadataImage,
+                groupMetadataManager,
+                offsetMetadataMaxSize
+            );
+        }
+    }
+
+    /**
+     * The logger.
+     */
+    private final Logger log;
+
+    /**
+     * The snapshot registry.
+     */
+    private final SnapshotRegistry snapshotRegistry;
+
+    /**
+     * The system time.
+     */
+    private final Time time;
+
+    /**
+     * The metadata image.
+     */
+    private MetadataImage metadataImage;
+
+    /**
+     * The group metadata manager.
+     */
+    private final GroupMetadataManager groupMetadataManager;
+
+    /**
+     * The maximum allowed metadata for any offset commit.
+     */
+    private final int offsetMetadataMaxSize;
+
+    /**
+     * The offsets keyed by topic-partition and group id.
+     */
+    private final TimelineHashMap<String, TimelineHashMap<TopicPartition, OffsetAndMetadata>> offsetsByGroup;
+
+    OffsetMetadataManager(
+        SnapshotRegistry snapshotRegistry,
+        LogContext logContext,
+        Time time,
+        MetadataImage metadataImage,
+        GroupMetadataManager groupMetadataManager,
+        int offsetMetadataMaxSize
+    ) {
+        this.snapshotRegistry = snapshotRegistry;
+        this.log = logContext.logger(OffsetMetadataManager.class);
+        this.time = time;
+        this.metadataImage = metadataImage;
+        this.groupMetadataManager = groupMetadataManager;
+        this.offsetMetadataMaxSize = offsetMetadataMaxSize;
+        this.offsetsByGroup = new TimelineHashMap<>(snapshotRegistry, 0);
+    }
+
+    /**
+     * Validates an OffsetCommit request.
+     *
+     * @param context The request context.
+     * @param request The actual request.
+     */
+    private void validateOffsetCommit(
+        RequestContext context,
+        OffsetCommitRequestData request
+    ) throws ApiException {
+        Group group;
+        try {
+            group = groupMetadataManager.group(request.groupId());
+        } catch (GroupIdNotFoundException ex) {
+            if (request.generationIdOrMemberEpoch() < 0) {
+                // If the group does not exist and generation id is -1, the request comes from
+                // either the admin client or a consumer which does not use the group management
+                // facility. In this case, a so-called simple group is created and the request
+                // is accepted.
+                group = groupMetadataManager.getOrMaybeCreateGenericGroup(request.groupId(), true);
+            } else {
+                // Maintain backward compatibility. This is a bit weird in the
+                // context of the new protocol though.
+                throw Errors.ILLEGAL_GENERATION.exception();
```

Comment:
I am considering whether we should allow `GROUP_ID_NOT_FOUND` in version >= 9 and keep using `ILLEGAL_GENERATION` for older versions. What do you think?

Reply:
I think it depends on the context. Is illegal generation implying that we are trying to create the group with an invalid generation ID -- but given we are using the new protocol (and not using generation id) and know that we are instead referring to an epoch of the group that group id not found makes more sense?

Reply:
Usually, the group should already exist when the consumer commits offsets, except in the "simple group" case discussed earlier. So if a member comes with a generation >= 0, the group should exist. I suppose that we return ILLEGAL_GENERATION because this is also correct. The generation is also incorrect if the group does not exist.

However, this ILLEGAL_GENERATION error does not make any sense for the new protocol as, as you said, we use the member epoch. Unfortunately, we cannot return STALE_MEMBER_EPOCH (the equivalent of ILLEGAL_GENERATION) because we don't know the group type if does not exist. This is why I am thinking about using GROUP_ID_NOT_FOUND for both cases. I have to check whether GROUP_ID_NOT_FOUND would be handled similarly to ILLEGAL_GENERATION though...

Reply:
If you agree, I will also update the KIP and notify vote thread.

Reply:
I also discussed this change with Jason and he agrees with this.

Reply:
Sorry for the delay -- this makes sense to me if the error handling is correct.

Reply:
I will do the client side changes separately.